mesa_frames/abstract/space.py

```diff
@@ -321,3 +383,345 @@ def random(self) -> Generator:
+
+        Parameters
+        ----------
+        pos : GridCoordinate | GridCoordinates
+            The positions to check for
+
+        Returns
+        -------
+        DataFrame
+            A dataframe with positions and a boolean column "full"
+        """
+        df = self._df_constructor(data=pos, columns=self._cells_col_names)
+        return self._df_add_columns(
+            df, ["full"], self._df_get_bool_mask(df, mask=self.full_cells, negate=True)
+        )
+
+    def move_to_empty(
+        self,
+        agents: IdsLike | AgentContainer | Collection[AgentContainer],
+        inplace: bool = True,
+    ) -> Self:
+        return self._move_agents_to_cells(agents, cell_type="empty", inplace=inplace)
+
+    def move_to_free(
+        self,
+        agents: IdsLike | AgentContainer | Collection[AgentContainer],
+        inplace: bool = True,
+    ) -> Self:
+        """Move agents to free cells/positions in the space (cells/positions where there is at least one spot available).
+
+        Parameters
+        ----------
+        agents : IdsLike | AgentContainer | Collection[AgentContainer]
+            The agents to move to free cells/positions
+        inplace : bool, optional
+            Whether to perform the operation inplace, by default True
+
+        Returns
+        -------
+        Self
+        """
+        return self._move_agents_to_cells(agents, cell_type="free", inplace=inplace)
+
+    def sample_cells(
+        self,
+        n: int,
+        cell_type: Literal["any", "empty", "free", "full"] = "any",
+        with_replacement: bool = True,
+    ) -> DataFrame:
+        """Sample cells from the grid according to the specified cell_type.
+
+        Parameters
+        ----------
+        n : int
+            The number of cells to sample
+        cell_type : Literal["any", "empty", "free", "full"], optional
+            The type of cells to sample, by default "any"
+        with_replacement : bool, optional
+            If the sampling should be with replacement, by default True
+
+        Returns
+        -------
+        DataFrame
+            A DataFrame with the sampled cells
+        """
+        match cell_type:
+            case "any":
+                condition = self._any_cell_condition
+            case "empty":
+                condition = self._empty_cell_condition
+            case "free":
+                condition = self._free_cell_condition
+            case "full":
+                condition = self._full_cell_condition
+        return self._sample_cells(n, with_replacement, condition=condition)
+
+    @abstractmethod
+    def get_neighborhood(
+        self,
+        radius: int | float | Sequence[int] | Sequence[float],
+        pos: DiscreteCoordinate | DataFrame | None = None,
+        agents: int | Sequence[int] | None = None,
+        include_center: bool = False,
+    ) -> DataFrame:
+        """Get the neighborhood cells from a given position.
```

Comment:
GIven that this is getting a neighborhood from just a position, the length of the output will often be small, for a vision of 1-5. As such, doing it as a DF would be slow. DF operations on a small number of rows are much slower than a Python loop.

It would be faster to get neighborhood in a bulk -- in lots of positions, but I'm not sure about this being intuitive.

Reply:
Sorry I updated the concrete version without updating the abstract one! Yes the idea is that you get neighborhood (possibly also of different length depending on the sequence of radius) for multiple positions/agents at once, in a single DF. This is to incentivize as much as possible the use of vectorized operations. Loops should almost never be used (apart from the step).